src/core/qgsaction.cpp

```diff
@@ -74,6 +87,183 @@ void QgsAction::run( const QgsExpressionContext &expressionContext ) const
+              {
+                bool escaped = false;
+                while ( pos != std::string::npos && header[pos - 1] == '\\' )
+                {
+                  pos = header.find( q2, pos + 1 );
+                  escaped = true;
+                }
+                ascii = header.substr( len, pos - len );
+                if ( escaped )
+                {
+                  std::string cleaned;
+                  for ( size_t i = 0; i < ascii.size(); ++i )
+                  {
+                    if ( ascii[i] == '\\' )
+                    {
+                      if ( i > 0 && ascii[i - 1] == '\\' )
+                      {
+                        cleaned.push_back( ascii[i] );
+                      }
+                    }
+                    else
+                    {
+                      cleaned.push_back( ascii[i] );
+                    }
+                  }
+                  ascii = cleaned;
+                }
+              }
+            }
+
+            std::string utf8;
+
+            const std::string u { R"(UTF-8'')" };
+            if ( const auto pos = header.find( u ); pos != std::string::npos )
+            {
+              utf8 = header.substr( pos + u.size() );
+            }
+
+            // Prefer ascii over utf8
+            if ( ascii.empty() )
+            {
+              if ( ! utf8.empty( ) )
+              {
+                filename = QString::fromStdString( utf8 );
+              }
+            }
+            else
+            {
+              filename = QString::fromStdString( ascii );
+            }
+          }
+          else if ( !reply->header( QNetworkRequest::KnownHeaders::ContentTypeHeader ).isNull() )
+          {
+            QString contentTypeHeader { reply->header( QNetworkRequest::KnownHeaders::ContentTypeHeader ).toString() };
+            // Strip charset if any
+            if ( contentTypeHeader.contains( ';' ) )
+            {
+              contentTypeHeader = contentTypeHeader.left( contentTypeHeader.indexOf( ';' ) );
+            }
+
+            QMimeType mimeType { QMimeDatabase().mimeTypeForName( contentTypeHeader ) };
+            if ( mimeType.isValid() )
+            {
+              filename = QStringLiteral( "download.%1" ).arg( mimeType.preferredSuffix() );
+            }
+          }
+
+          QTemporaryDir tempDir;
+          tempDir.setAutoRemove( false );
+          tempDir.path();
+          const QString tempFilePath{ tempDir.path() + QDir::separator() + filename };
+          QFile tempFile{ tempFilePath };
+          tempFile.open( QIODevice::WriteOnly );
+          tempFile.write( replyData );
+          tempFile.close();
+          QDesktopServices::openUrl( QUrl::fromLocalFile( tempFilePath ) );
+        }
+        else
+        {
+          QgsMessageLog::logMessage( QObject::tr( "Redirect is not supported!" ), QStringLiteral( "Form Submit Action" ), Qgis::MessageLevel::Critical );
```

Comment:
Could we just use `req.setAttribute(QNetworkRequest::FollowRedirectsAttribute, true);` to delegate handling of this to Qt?

Reply:
We could, but redirect on POST is not a common HTTP workflow (it has potential security issues and browsers should prompt the user for confirmation), I'm not sure it is a wise move to support it.

Reply:
Fair enough!